docs/madr/decisions/006-mesh-traffic-permission.md

```diff
@@ -0,0 +1,576 @@
+# MeshTrafficPermission
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4222
+
+## Context and Problem Statement
+
+[New policy matching MADR](./005-policy-matching.md) introduces a new approach how Kuma applies configuration to proxies.
+Rolling out strategy implies creating a new set of policies that satisfy a new policy matching.
+Current MADR aims to define a MeshTrafficPermission.
+
+## Considered Options
+
+* Create a MeshTrafficPermission policy
+
+## Decision Outcome
+
+Chosen option: create a MeshTrafficPermission policy
+
+### Overview
+
+MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
+There are might be several reasons why it could be useful to block unwanted traffic:
+* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
+* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)
+
+### Specification
+
+Top-level targetRef can have all available kinds:
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+  name: ...
+```
+
+MeshTrafficPermission is an inbound policy that's why it has a "from" array:
+
+```yaml
+from:
+  - targetRef:
+      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+      name: ...
+```
+
+Configuration for MeshTrafficPermission has 2 fields "action" and "shadow_action" with values "ALLOW" or "DENY".
+So the overall specification looks like this:
+
+```yaml
+type: MeshTrafficPermission
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+    name: ...
+  from:
+    - targetRef:
+        kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+        name: ...
+      default:
+        action: ALLOW|DENY
+        shadow_action: ALLOW|DENY
+```
+
+We don't want to mix mTLS configs (like permissive mTLS or TLS version) with MeshTrafficPermission. 
+The main reason is that mTLS settings are applied for the entire inbound, while permissions could be set for a group of clients. 
+
+### Shadow actions
+
+Shadow actions are not enforced but emit stats and can be used for rule testing.
+For example, before creating "action: DENY" rule, user creates "shadow_action: DENY" rule and checks the stats.
+If there are no unexpected denied requests from the clients then it's safe to change "shadow_action" to "action".
+
+Alternatively we could introduce `shadow: true|false` instead of `shadow_action: ALLOW|DENY`. 
+But this approach has its flaws:
+* policy matching makes it less obvious with merging. It's easy to create a rule like:
+    ```yaml
+    targetRef:
+       kind: Mesh
+    default:
+      action: DENY
+      shadow: true
+   ```
+   and unintentionally shadow all less specific rules.
```

Comment:
How is it easier to do this vs 

```yaml
targetRef:
  kind: Mesh
default:
  shadowAction: DENY
```
?

Reply:
because you won't affect `action: DENY` of the less specific rules. But with `shadow: true` you can affect `action` field of other rules (because fields are correlated)

Reply:
That's only because we previously defined a rigid rule about merging top level fields that applies to _all policies_. It's unintentionally confusing because we interpret the semantics as if it were:

```
default:
  rule:
    action: ALLOW|DENY
    shadow: true
```

Anyway, action/shadowAction can just as easily lead to unintentionally overriding less specific rules in exactly the same way, can't it?

=> it should be `action: ALLOW|DENY|SHADOW_DENY|SHADOW_ALLOW`.

Reply:
> Anyway, action/shadowAction can just as easily lead to unintentionally overriding less specific rules in exactly the same way, can't it?

That is to say, they're not orthogonal to one another either:

```
type: MeshTrafficPermission
name: global
spec:
  targetRef:
    kind: Mesh
  from:
    - targetRef:
         kind: MeshService
         name: backend
       default:
         shadowAction: ALLOW
---
type: MeshTrafficPermission
name: mtp-web
spec:
  targetRef:
    kind: MeshService
    name: web
  from:         
    - targetRef:
         kind: Mesh
       default:
         action: ALLOW
```

and now I have

```
action: ALLOW
shadowAction: ALLOW
```

when the user probably wanted:

```
action: DENY
shadowAction: ALLOW
```

They forgot to set `action: DENY` in the more specific policy just like they forgot to set `shadow: false` in the more specific policy in the `shadow: bool` example.

Reply:
Except it's possible to differentiate a field unset from a boolean unset (as it's false). So when you don't set a field lower down it uses whatever is inherited from the parent. This is impossible with booleans right?

Reply:
I'm not sure what you mean but `bool <=> {ALLOW, DENY}` and `*bool <=> *{ALLOW, DENY}`. There's no difference, is there?

Reply:
yeah, we could use `google.protobuf.BoolValue` to distinguish `bool` and `nil`. But I guess it doesn't matter since we don't want `shadow: true|false` anyway